### DISCUSSION TITLE
Removing duplicates when reading a file

### DISCUSSION THREAD

**Discussion by Jenix-Park** [melanchall/drywetmidi#222] [Q&A] [answered]
Hey, how have you been?

I wonder what's the simplest and fastest way to remove duplicate "DamperPedal" events when reading a random MIDI file?
Let's say we have a file with multiple sustain-on events in a row and one sustain-off event after them.
What I want are just one sustain-on and sustain-off.

Is there any way to simply achieve this at a low-performance cost?

**Comment by melanchall** [comment, marked as answer]
I think something like that should work:

```csharp
var flag = false;
midiFile.RemoveTimedEvents(e =>
{
    var midiEvent = e.Event;
    if (/* midiEvent is sustain on */)
    {
        if (flag)
            return true; // event should be removed

        flag = true;
        return false; // we're on a first sustain on
    }

    flag = false;
    return false;    
});
```
This code will remove all subsequent sustain on events. But I in fact don't know what "sustain on" event is. I see there is [controller type](https://melanchall.github.io/drywetmidi/api/Melanchall.DryWetMidi.Common.ControlName.html)

```
DamperPedal 	Damper Pedal On/Off (Sustain).
```

But it means both on and off states. So please implement condition here

```csharp
/* midiEvent is sustain on */
```

by yourself. Or clarify what event types do you mean. You can also send me a sample file pointing where you see duplicated events (and where sustain on and off ones). But I'll be able to look into the file a week later, I'm on vacation right now without access to my computer.

**Reply by Jenix-Park** [reply]
Thank you, have a good time, love you!

**Reply by melanchall** [reply]
Thank you!

**Reply by Jenix-Park** [reply]
@melanchall
Hello, how have you been? 🙂

I just checked the latest version and wanted to thank you for adding `SanitizingSettings.RemoveDuplicatedNotes`! :)

I was wondering if you've also added a similar feature for handling duplicate damper pedal events since last time.
I checked the manual but couldn’t find anything about it.

I’m currently using the workaround code you suggested earlier, so it’s not really needed.
However, for future projects, it would be great to have a built-in solution, as it would improve maintainability and readability.

Thanks a lot!

**Reply by melanchall** [reply]
@Jenix-Park I'm fine, thank you for kind words! Hope you're well too.

Hmm, as for damper pedal, I think it would be more correct to filter out duplicated _Control Change_ events in general. But we need to skip events related to RPNs/NRPNs. Sounds as a plan :-)

Thank you for the idea again! I'll provide a solution as soon as possible.

**Reply by Jenix-Park** [reply]
@melanchall Hey 😃
As I mentioned, I’m personally fine without it, but I just got curious after the new update, so I had to ask.
As the built-in sanitizer becomes more capable, it’ll definitely improve both the user experience and maintenance.
It’s great that you came up with another good idea, haha.

Cheers!

**Reply by melanchall** [reply]
@Jenix-Park I've published the [7.2.1-prerelease1](https://www.nuget.org/packages/Melanchall.DryWetMidi/7.2.1-prerelease1) version with the new `RemoveDuplicatedControlChangeEvents` property in the `SanitizingSettings`. If you want, you can use this prerelease version to solve your task :-)

Thank you for your feedback!

**Reply by Jenix-Park** [reply]
@melanchall Hey! 😊

Wow, that was really quick! Thank you so much for the update. While it wasn’t something urgent for my current project, I really appreciate you adding it. I’ll leave things as they are for now, but I’ll definitely consider updating to the new version when I have some time and for future projects.

Thanks again for your awesome responsiveness!

Cheers!

**Reply by melanchall** [reply]
:rocket: **8.0.0 version is released now!**

This version introduces [`SanitizingSettings.RemoveDuplicatedControlChangeEvents`](https://melanchall.github.io/drywetmidi/api/Melanchall.DryWetMidi.Tools.SanitizingSettings.html#Melanchall_DryWetMidi_Tools_SanitizingSettings_RemoveDuplicatedControlChangeEvents) property mentioned above. Thank you!

**Reply by Jenix-Park** [reply]
Good new! I'll try it! Thanks :)
